UI: Add preloading to stories highlighted in the sidebar

ndelangen opened the pull request:
## What I did

Optimizations:
- When you hover over a story in the sidebar we could sent a message to the preview the user is likely to go to that story.
- When a user navigates to a story using the keyboard shortcuts we could sent a message to the preview the user is likely to go to that story.

tmeasday commented on lib/ui/src/components/sidebar/Tree.tsx:
```diff
@@ -169,6 +170,11 @@ const Node = React.memo<NodeProps>(
               onSelectStoryId(item.id);
             }}
+            onMouseEnter={() => {
+              if (item.isLeaf) {
+                api.emit(PRELOAD_STORIES, [item.id]);
+              }
```
We want to emit the signal when you mouse over the component *before* you click on the first story (there's no point in preloading sibling stories once they are already visible, right?).

I think something like this will do it (factoring in docs-only stories too)?

```suggestion
              if (item.isComponent) {
                api.emit(PRELOAD_STORIES, [item.isLeaf ? item.id : item.children[0]]);
              }
```

tmeasday reviewed (approved):
I tried this out and found two problem:

1. When mousing over the sidebar, it only preloads at the story level, which is "too late" as a sibling story will already have been loaded. (similar when using the keyboard).

2. It doesn't do anything in the search panel. I thought we planned to do similar behaviour on mouse/keyboard selection?

tmeasday commented on lib/ui/src/components/sidebar/useHighlighted.ts:
```diff
@@ -109,6 +112,10 @@ export const useHighlighted = ({
         const didRunAround = isArrowUp ? nextIndex === highlightable.length - 1 : nextIndex === 0;
         highlightElement(highlightable[nextIndex], didRunAround);
+
+        if (highlightable[nextIndex].getAttribute('data-nodetype') === 'story') {
+          api.emit(PRELOAD_STORIES, [highlightedRef.current?.itemId]);
+        }
```
We need to do similar logic to the above `onMouseEnter` handler.

tmeasday reviewed (commented):
X

tmeasday reviewed (changes requested):
Meant to request changes

ndelangen commented:
@tmeasday I think I made all the requested changes

tmeasday commented on lib/preview-web/src/PreviewWeb.tsx:
```diff
@@ -209,8 +209,10 @@ export class PreviewWeb<TFramework extends AnyFramework> extends Preview<TFramew
-    await Promise.all(ids.map((id) => this.storyStore.loadStory({ storyId: id })));
+  async onPreloadStories({ ids }: { ids: string[] }) {
+    await Promise.all(
+      ids.map((id) => this.storyStore.loadStory({ storyId: id }).catch(() => undefined))
```
Is this still needed? Does the targetting work?

ndelangen replied:
@tmeasday so it's possible that we're trying to preload a story in a ref we haven't loaded the iframe for yet.

Because of the way the targeting works, if we can't find the targeted iframe, we'll use the currently active iframe.
Which can cause the event to be targeted to the wrong iframe, causing an error if the storyId does not exists there.

Therefore I opted to catch the error thrown.
(and try to target the correct iframe)

ndelangen replied:
> I did notice the target attribute doesn't seem consistent:

great catch, I think I fixed that

tmeasday replied:
OK, seems reasonable, can you make a comment about that?

yannbf replied:
Done!

tmeasday reviewed (changes requested):
Functionally it works great now.

I did notice the `target` attribute doesn't seem consistent:

<img width="728" alt="image" src="https://user-images.githubusercontent.com/132554/165415183-affa4f08-673a-4aca-b40d-e31698f9bcc0.png">

<img width="586" alt="image" src="https://user-images.githubusercontent.com/132554/165415203-ab6d2a6f-34e4-4b14-8a04-6c0400e5927f.png">

Also, I wonder if we should just add an `api.preloadStory` method to the `stories` module to avoid repeating the logic all over the place.

yannbf commented on lib/ui/src/components/sidebar/SearchResults.tsx:
```diff
@@ -127,6 +129,26 @@ const Result: FunctionComponent<
+        );
+      }
+    }
+  }, [props.isHighlighted]);
```
Shouldn't the item be a dependency too?

```suggestion
  }, [props.isHighlighted, item]);
```

yannbf commented on lib/ui/src/components/sidebar/SearchResults.tsx:
```diff
@@ -203,6 +226,26 @@ export const SearchResults: FunctionComponent<{
     }, [enableShortcuts, isLoading]);
 
+    const mouseOverHandler = useCallback((event: MouseEvent) => {
+      // @ts-ignore
+      const storyId = event.currentTarget.getAttribute('data-id');
+      // @ts-ignore
+      const refId = event.currentTarget.getAttribute('data-refid');
```
```suggestion
      const currentTarget = event.currentTarget as HTMLElement;
      const storyId = currentTarget.getAttribute('data-id');
      const refId = currentTarget.getAttribute('data-refid');
```

yannbf reviewed (commented):
I tested in different projects using webpack4 and 5 and the behavior is super awesome!

There was one scenario that didn't go really well (kind of expected to be honest @tmeasday) which is when using LazyCompilation. If the user hovers over +10 stories fast, then webpack goes haywire:

https://user-images.githubusercontent.com/1671563/167392716-b0a76215-08a3-4232-b8a1-1901685c18ae.mov

ndelangen commented:
Should report as a bug to webpack I guess.

I propose we add a debounce here:
https://github.com/storybookjs/storybook/pull/17964/files#diff-1c750d3374a0e928fde974aae523e9257f6b6c23321e70527841850683609419R212-R226

tmeasday commented:
Ok! Wanna add it in this PR or another next cycle?

tmeasday commented on lib/ui/src/components/sidebar/SearchResults.tsx:
```diff
@@ -127,6 +129,26 @@ const Result: FunctionComponent<
+  useEffect(() => {
+    if (props.isHighlighted) {
+      if (item.isComponent) {
+        api.emit(
+          PRELOAD_STORIES,
+          { ids: [item.isLeaf ? item.id : item.children[0]] },
+          {
+            options: {
+              target:
+                item.refId && item.refId !== 'storybook_internal'
+                  ? `storybook-ref-${item.refId}`
+                  : 'storybook-preview-iframe',
+            },
+          }
```
We have this same code in 4 places, can we add `api.preloadStories` to `modules/stories.ts` 🙏 ?

tmeasday reviewed (changes requested):
@ndelangen there seems to be some stories failing

tmeasday commented on lib/preview-web/src/PreviewWeb.tsx:
```diff
@@ -234,8 +235,20 @@ export class PreviewWeb<TFramework extends AnyFramework> extends Preview<TFramew
+  async onPreloadStories({ ids }: { ids: string[] }) {
+    await Promise.all(
+      ids.map((id) =>
+        this.storyStore.loadStory({ storyId: id }).catch(() => {
```
Can we use `loadEntry` here?

ndelangen replied:
✅

tmeasday commented on lib/ui/src/components/preview/preview.tsx:
```diff
@@ -165,7 +165,10 @@ const Preview = React.memo<PreviewProps>((props) => {
             viewMode,
             options: {
-              target: refId ? `storybook-ref-${refId}` : 'storybook-preview-iframe',
+              target:
+                refId && refId !== 'storybook_internal'
+                  ? `storybook-ref-${refId}`
+                  : 'storybook-preview-iframe',
```
This targetting code seems to be repeated a bunch of times, maybe we can make `api.emit()` do it internally?

tmeasday commented on lib/ui/src/components/sidebar/Tree.tsx:
```diff
@@ -243,6 +246,19 @@ const Node = React.memo<NodeProps>(
         }}
+        onMouseEnter={() => {
+          if (item.isComponent) {
+            api.emit(PRELOAD_STORIES, {
+              ids: [item.children[0]],
+              options: {
+                target:
+                  refId && refId !== 'storybook_internal'
+                    ? `storybook-ref-${refId}`
+                    : 'storybook-preview-iframe',
+              },
+            });
+          }
```
Can we add `api.preloadEntries()` to the stories module? Also we could consider renaming the event to `PRELOAD_ENTRIES` as they don't necessarily need to be stories (they can be docs)

ndelangen replied:
I renamed the event-name..

Are you suggesting changing this:
```ts
            api.emit(PRELOAD_ENTRIES, {
              ids: [item.isLeaf ? item.id : item.children[0]],
              options: { target: refId },
            });
```
to:
```ts
            api.preloadEntries({
              ids: [item.isLeaf ? item.id : item.children[0]],
              options: { target: refId },
            });
```
?

tmeasday replied:
If that's all it's doing then I guess I agree with your (implied) point that it's not worth it!

tmeasday reviewed (approved):
LGTM

ndelangen committed:
rename PRELOAD_STORIES to PRELOAD_ENTRIES event
auto map the ref id in lib/api when calling emit
emit preload event when hovering over sidebar items or selecting them with the keyboard

ndelangen committed:
fix
